{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/anaconda/lib/python3.6/site-packages/numba/errors.py:131: UserWarning: Insufficiently recent colorama version found. Numba requires colorama >= 0.3.9\n",
      "  warnings.warn(msg)\n"
     ]
    }
   ],
   "source": [
    "import subprocess,os,copy,pickle,re,time,itertools,random\n",
    "import numpy as np\n",
    "import statsmodels.api as sm\n",
    "import math as ma\n",
    "import pandas as pd\n",
    "import celldiscoveryutilities as utils\n",
    "# Import 3rd party packages\n",
    "from sklearn import  metrics, neighbors\n",
    "from sklearn.decomposition import PCA\n",
    "from sklearn.cluster import KMeans\n",
    "from sklearn.preprocessing import scale\n",
    "from sklearn.manifold import TSNE\n",
    "from sklearn.utils.validation import check_array\n",
    "from scipy.stats import poisson, binom\n",
    "from numpy.random import beta, poisson\n",
    "import multiprocessing as mp\n",
    "import matplotlib.patches as mpatches\n",
    "import seaborn as sns\n",
    "from scipy import stats\n",
    "import matplotlib.pyplot as plt\n",
    "from scipy.sparse import issparse\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "#All the R related scripts\n",
    "Rcmd='/usr/local/bin/Rscript'\n",
    "var_genes_script='/mnt/kauffman/nbserver/notebooks/users/mtakai/R/test.R'\n",
    "deseq_norm_script='/mnt/kauffman/nbserver/notebooks/users/mtakai/R/get_deseq_normalization.R'\n",
    "infname='../results/hs_scores.tab'\n",
    "hs_df=utils.read_inputdata(infname)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "#Globala_variables\n",
    "PROTOCOL_PALETTE_DICT={'SCRBseq': (0.00784313725490196, 0.24313725490196078, 1.0),\n",
    "                   'Smartseq2': (1.0, 0.48627450980392156, 0.0),\n",
    "                   'CELseq': (0.10196078431372549, 0.788235294117647, 0.2196078431372549),\n",
    "                   'MARSseq': (0.9098039215686274, 0.0, 0.043137254901960784),\n",
    "                   'DropSeq': (0.5450980392156862, 0.16862745098039217, 0.8862745098039215),\n",
    "                   'Smartseq': (0.6235294117647059, 0.2823529411764706, 0.0)}\n",
    "\n",
    "PROTOCOL_BATCHES_PALETTE_DICT={'SCRBseqA': (0.00784313725490196, 0.24313725490196078, 1.0),\n",
    "                       'SCRBseqB': (0.00784313725490196, 0.729411764705882, 1.0),\n",
    "                       'SmartSeq2A': (1.0, 0.48627450980392156, 0.0),\n",
    "                       'SmartSeq2B': (1.0, 0.972549019607843, 0.0),\n",
    "                       'CELseq2A': (0.10196078431372549, 0.788235294117647, 0.2196078431372549),\n",
    "                       'CELseq2B': (0.10196078431372549, 1.0, 0.2196078431372549),\n",
    "                       'MARSseqA': (0.9098039215686274, 0.0, 0.043137254901960784),\n",
    "                       'MARSseqB': (0.9098039215686274, 0.5, 0.043137254901960784),\n",
    "                       'DropSeqA': (0.5450980392156862, 0.16862745098039217, 0.8862745098039215),\n",
    "                       'DropSeqB': (0.5450980392156862, 0.7, 0.8862745098039215),\n",
    "                       'SmartSeqA': (0.6235294117647059, 0.2823529411764706, 0.0),\n",
    "                       'SmartSeqB': (0.6235294117647059, 0.56, 0.0)}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def test_method(DGE,bins=20,verbose=0):\n",
    "    \"\"\"Break expression levels per bin\"\"\"\n",
    "    DGE_perm=DGE.copy()\n",
    "    GSUMS=np.sum(DGE,axis=1)\n",
    "    breakvec = np.linspace(1,100,bins)\n",
    "    breaks=[]\n",
    "    for breaker in breakvec:\n",
    "        breaks.append(np.percentile(GSUMS,breaker))\n",
    "    breaks=np.unique(breaks)\n",
    "    for i in range(len(breaks)-1):\n",
    "        if verbose==1:\n",
    "            print(np.round((1.0*i)/(len(breaks)-1)))\n",
    "        for j in range(len(DGE.columns)):\n",
    "            curlogical=np.logical_and(GSUMS>breaks[i],GSUMS<=breaks[i+1])\n",
    "            DGE_perm.ix[curlogical,j]=np.random.permutation(DGE_perm.ix[curlogical,j])\n",
    "    return DGE_perm\n",
    "\n",
    "## Step 1. Functions for reading,transformation and perturbing gene expression dataframe\n",
    "def read_inputdata(filepath,sep='\\t',header=0,index_col=0,nrows=None):\n",
    "    \n",
    "    '''Reads data table into a data frame'''\n",
    "    \n",
    "    with open(filepath,'r') as infh:\n",
    "        \n",
    "        input_df=pd.read_csv(filepath,sep=sep,header=header,index_col=index_col,nrows=nrows)\n",
    "        return (input_df)\n",
    "    \n",
    "def prepare_data(inputdata, fc, no_genes,  no_cells, method='multiplicative',quantiles=None,\n",
    "                 filter_genes_pattern=None,genes_and_cells_cut_off=None,normalize=None,\n",
    "                 pseudo_value=1,find_var_genes=None,preselected_genes=None,\n",
    "                 gene_modification_profile_df=None,gene_parameters_df=None,genes_in_all_cells=False):\n",
    "    \n",
    "    data = remove_nonexpressedgenes(inputdata)\n",
    "    \n",
    "    if filter_genes_pattern:\n",
    "        #Useful in cases where user may want to exclude e.g ERCC\n",
    "        data = remove_genes(inputdata=data, gene_pattern=filter_genes_pattern)\n",
    "    if normalize:\n",
    "        #In case you provide counts instead of normalized expression\n",
    "        norm_data_dict=utils.get_genes_sf(counts_df=data,pseudo_value=pseudo_value,loc_func=np.median,\n",
    "                                          return_norm_expr=True)\n",
    "        data =norm_data_dict['norm']\n",
    "    if quantiles:\n",
    "        #Get only genes from a specific mean quantile interval\n",
    "        data=genes_quantiles(data,lower_quantile=quantiles[0],upper_quantile=quantiles[1]) \n",
    "    if genes_and_cells_cut_off:\n",
    "        #Gets genes expressed in a specific number of cells\n",
    "        data=filter_genes_and_samples(data,cut_off=genes_and_cells_cut_off[0],\n",
    "                                      no_samples=genes_and_cells_cut_off[1],\n",
    "                                      genes_count_cut_off=genes_and_cells_cut_off[2])\n",
    "        \n",
    "    perturbed_data =perturb_data(inputdata=data,fc=fc,no_genes=no_genes,no_cells=no_cells,\n",
    "                                 method=method,preselected_genes=preselected_genes,\n",
    "                                 genes_in_all_cells=genes_in_all_cells,\n",
    "                                 gene_modification_profile_df=gene_modification_profile_df,\n",
    "                                 gene_parameters_df=gene_parameters_df)\n",
    "    \n",
    "    return perturbed_data\n",
    "\n",
    "def perturb_data(inputdata,fc,no_genes,no_cells,method='multiplicative',preselected_genes=None,\n",
    "                 genes_in_all_cells=False,gene_modification_profile_df=None,gene_parameters_df=None):\n",
    "    '''Modifys expression data frame of a given set of genes, in target cells using the input method''' \n",
    "    all_cells=inputdata.columns.tolist()\n",
    "    #np.random.seed(seed=115678)\n",
    "    cells=np.random.choice(all_cells,no_cells,replace=False)\n",
    "    all_genes_set=set(inputdata.index.tolist())\n",
    "    #np.random.seed(seed=None)\n",
    "    genes_to_select_from_set=all_genes_set\n",
    "    if preselected_genes:\n",
    "        genes_to_select_from_set=genes_to_select_from_set.intersection(preselected_genes)\n",
    "    \n",
    "    if genes_in_all_cells == True:\n",
    "        temp_target_cell_df=inputdata.loc[:,cells]\n",
    "        temp_target_cells_genes_list=temp_target_cell_df.index[temp_target_cell_df[temp_target_cell_df>0.0].count(axis=1)\n",
    "                                                            ==no_cells].tolist()\n",
    "        \n",
    "        genes_to_select_from_set=genes_to_select_from_set.intersection(temp_target_cells_genes_list)\n",
    "    try:\n",
    "        out_dict = {}\n",
    "        \n",
    "        if (method == 'multiplicative'):\n",
    "            genes = np.random.choice(list(genes_to_select_from_set), size=no_genes, replace=False)\n",
    "            out_dict['genes']= genes\n",
    "            out_dict['cells']= cells\n",
    "            out_dict['fc']= fc\n",
    "            mod_data = multiplicative_modification(inputdata=inputdata, cells=cells, genes=genes, fc=fc)\n",
    "            out_dict['data']=mod_data\n",
    "        if (method == 'mean'):\n",
    "            genes = np.random.choice(list(genes_to_select_from_set), size=no_genes, replace=False)\n",
    "            out_dict['genes'] = genes\n",
    "            out_dict['cells'] = cells\n",
    "            out_dict['fc'] = fc\n",
    "            mod_data = mean_modification(inputdata=inputdata, genes=genes, cells=cells, fc=fc)\n",
    "            out_dict['data']=mod_data\n",
    "            \n",
    "        if (method == 'synthetic'):\n",
    "            genes = np.random.choice(list(genes_to_select_from_set), size=no_genes, replace=False)\n",
    "            out_dict['genes'] = genes\n",
    "            out_dict['cells'] = cells\n",
    "            out_dict['fc'] = fc\n",
    "            mod_data = add_synthetic_genes_modification(inputdata=inputdata, genes=genes, cells=cells, fc=fc)\n",
    "            out_dict['data']=mod_data\n",
    "            \n",
    "        if (method=='multiplicative_modification_space'):\n",
    "            mod_dict=modification_in_subspace(inputdata=inputdata,cells=cells,no_genes=no_genes,fc=fc,\n",
    "                                              gene_parameters_df=gene_parameters_df,all_genes_set=all_genes_set,\n",
    "                                              genes_to_select_from_set=genes_to_select_from_set,\n",
    "                                              gene_modification_profile_df=gene_modification_profile_df)\n",
    "            out_dict['cells'] = cells\n",
    "            out_dict['fc'] = fc\n",
    "            out_dict['data'] = mod_dict['data']\n",
    "            out_dict['genes'] = mod_dict['genes']\n",
    "            out_dict['matching_genes'] = mod_dict['matching_genes']\n",
    "            \n",
    "        if (method=='add_marker_gene_modification'):\n",
    "            mod_dict=add_marker_gene_modification(inputdata=inputdata,cells=cells,no_genes=no_genes,fc=fc,\n",
    "                                              gene_parameters_df=gene_parameters_df,all_genes_set=all_genes_set,\n",
    "                                              genes_to_select_from_set=genes_to_select_from_set,\n",
    "                                              gene_modification_profile_df=gene_modification_profile_df)\n",
    "            out_dict['cells'] = cells\n",
    "            out_dict['fc'] = fc\n",
    "            out_dict['data'] = mod_dict['data']\n",
    "            out_dict['genes'] = mod_dict['genes']\n",
    "            out_dict['matching_genes'] = mod_dict['matching_genes'] \n",
    "        return out_dict\n",
    "    \n",
    "    \n",
    "    except KeyboardInterrupt as key_err:\n",
    "        print(key_err, 'thrown!!!', 'keyboard interruption during modification')\n",
    "        \n",
    "        pass\n",
    "    except ValueError as val_err:\n",
    "        print(val_err,'ValueError error captured during the modification')\n",
    "        pass\n",
    "    except IndexError as ind_error:\n",
    "        print(ind_error,'IndexError error captured during modification')\n",
    "        \n",
    "        pass\n",
    "    \n",
    "    pass\n",
    "\n",
    "def genes_quantiles(inputdata,lower_quantile=0.0,upper_quantile=1.0):\n",
    "    \n",
    "    # returns the input data frame after selecting gene from the specific mean-quantiles \n",
    "    \n",
    "    mean_series=inputdata.mean(axis=1)\n",
    "    lower_cut_off=mean_series.quantile(np.float(lower_quantile))\n",
    "    upper_cut_off = mean_series.quantile(np.float(upper_quantile))\n",
    "    \n",
    "    mean_series_filt=mean_series[mean_series>=lower_cut_off]\n",
    "    \n",
    "    if np.round(upper_quantile,2)<1.00:\n",
    "        \n",
    "        mean_series_filt=mean_series_filt[mean_series_filt < upper_cut_off]  \n",
    "        \n",
    "    else:\n",
    "        mean_series_filt[mean_series_filt <= upper_cut_off]\n",
    "        \n",
    "    output_data=inputdata.loc[mean_series_filt.index,inputdata.columns]\n",
    "    return output_data\n",
    "\n",
    "def multiplicative_modification(inputdata,genes,cells,fc=2,mean_interval=2.0,drop_out_interval=0.1,cv2=None):\n",
    "    \n",
    "    '''Given specific samples and genes, modifys the expression levels based on the provided fold change'''\n",
    "   \n",
    "    inputdata.loc[genes, cells]= inputdata.loc[genes, cells].mul(fc)\n",
    "    \n",
    "    return inputdata\n",
    "\n",
    "def modification_in_subspace(inputdata,cells,no_genes,fc,gene_parameters_df,all_genes_set=None,genes_to_select_from_set=None,\n",
    "                            gene_modification_profile_df=None):\n",
    "    \n",
    "    '''Given specific cells and number of genes, modifys the expression levels based on the provided fold change and\n",
    "    ensures the mean is within a target subspace '''\n",
    "    all_genes_set=set(inputdata.index.tolist()) if all_genes_set is None else all_genes_set.intersection(inputdata.index.tolist())\n",
    "    profiled_fc=None if gene_modification_profile_df is None else gene_modification_profile_df.fc.unique().tolist()\n",
    "    genes_to_select_from_set=all_genes_set if genes_to_select_from_set is None else genes_to_select_from_set.intersection(all_genes_set)\n",
    "    gene_parameters_df=gene_parameters_df.loc[all_genes_set,:]\n",
    "    all_genes_set_as_list=list(all_genes_set)\n",
    "    len_cells=len(cells)\n",
    "    all_cells=inputdata.columns.tolist()     \n",
    "    genes=[]      \n",
    "    matching_genes=[]\n",
    "    matching_genes_dict={}\n",
    "    out_dict={}\n",
    "    matched_cells=np.random.choice(all_cells,len_cells,replace=False)\n",
    "    if fc==0.0:\n",
    "        genes_to_select_from_temp_list=list(genes_to_select_from_set)\n",
    "        selected_genes=np.random.choice(genes_to_select_from_temp_list,no_genes,replace=False).tolist()\n",
    "        out_dict['genes'] = selected_genes\n",
    "        out_dict['cells'] = cells\n",
    "        out_dict['fc'] = fc\n",
    "        out_dict['data'] = inputdata\n",
    "        out_dict['matching_genes']=selected_genes    \n",
    "    elif profiled_fc is None or fc not in profiled_fc:\n",
    "        mean_series = gene_parameters_df.mean_expr\n",
    "        genes_to_select_from_temp_list=list(genes_to_select_from_set)\n",
    "        len_genes_to_select_from_temp_list=len(genes_to_select_from_temp_list)\n",
    "        len_limit=no_genes if no_genes<=len_genes_to_select_from_temp_list else len_genes_to_select_from_temp_list\n",
    "        checked_genes = []\n",
    "        matched_genes_tracker=set()\n",
    "        while len(matching_genes_dict)<=len_limit:\n",
    "            temp_gns_to_check=list(genes_to_select_from_set.difference(checked_genes))\n",
    "            if len(temp_gns_to_check)==0:break\n",
    "            temp_gn=np.random.choice(temp_gns_to_check, 1)[0]\n",
    "            checked_genes.append(temp_gn)\n",
    "            target_mean = np.float(mean_series.loc[temp_gn])*fc\n",
    "            interval=0.1*target_mean\n",
    "            mod_mean_interval = [target_mean - interval, target_mean+ interval]\n",
    "            temp_matching_genes_list = gene_parameters_df[(gene_parameters_df.mean_expr >= mod_mean_interval[0]) &\n",
    "                                                          (gene_parameters_df.mean_expr <= mod_mean_interval[1])].index.tolist()\n",
    "            if not temp_matching_genes_list: continue\n",
    "            #np.random.seed(seed=1111)\n",
    "            #matching_genes_list=list(all_genes_set.intersection(temp_matching_genes_list).intersection(tmp_mod_sig_var_df.index))\n",
    "            matching_genes_set=all_genes_set.intersection(temp_matching_genes_list)\n",
    "            if not matching_genes_set : continue\n",
    "            diff_matching_genes_set=matching_genes_set.difference(matched_genes_tracker)\n",
    "            #matching_genes_set_as_list=\n",
    "            tmp_matching_gene=np.random.choice(list(matching_genes_set),1)[0] if diff_matching_genes_set else np.random.choice(list(matching_genes_set),1)[0]\n",
    "            genes.append(temp_gn)\n",
    "            matching_genes.append(tmp_matching_gene)\n",
    "            matching_genes_dict[temp_gn]=tmp_matching_gene\n",
    "            matched_genes_tracker.add(tmp_matching_gene)\n",
    "        del matching_genes_dict\n",
    "        del genes_to_select_from_set\n",
    "        \n",
    "        del mean_series\n",
    "        \n",
    "        del all_genes_set\n",
    "        \n",
    "        del gene_parameters_df\n",
    "            \n",
    "        del profiled_fc\n",
    "            \n",
    "        del checked_genes \n",
    "        \n",
    "        del matched_genes_tracker\n",
    "\n",
    "        inputdata.loc[genes,cells]=np.array(inputdata.loc[matching_genes,matched_cells])\n",
    "        out_dict['genes'] = genes\n",
    "        out_dict['cells'] = cells\n",
    "        out_dict['fc'] = fc\n",
    "        out_dict['data'] = inputdata\n",
    "        out_dict['matching_genes']=matching_genes\n",
    "    else:\n",
    "        gene_modification_profile_df=gene_modification_profile_df.astype({'genes':str})\n",
    "        temp_gene_modification_profile_df = gene_modification_profile_df[(gene_modification_profile_df.fc == fc)\n",
    "                                                                         & (gene_modification_profile_df.genes!='nan')]\n",
    "        genes_to_select_from_set = genes_to_select_from_set.intersection(temp_gene_modification_profile_df.gene.tolist())\n",
    "        genes_to_select_from_temp_list=list(genes_to_select_from_set)\n",
    "        len_genes_to_select_from_temp_list=len(genes_to_select_from_temp_list)\n",
    "        len_limit=no_genes if no_genes<=len_genes_to_select_from_temp_list else len_genes_to_select_from_temp_list\n",
    "        checked_genes=[]\n",
    "        matched_genes_tracker=set()\n",
    "        while len(matching_genes_dict) < len_limit:\n",
    "            \n",
    "            temp_gns_to_check=list(genes_to_select_from_set.difference(checked_genes))\n",
    "            \n",
    "            if len(temp_gns_to_check)==0:break\n",
    "            \n",
    "            temp_gn=np.random.choice(temp_gns_to_check, 1)[0]\n",
    "            \n",
    "            checked_genes.append(temp_gn)\n",
    "            \n",
    "            matching_genes_list = [temp_gene_modification_profile_df[temp_gene_modification_profile_df.gene == temp_gn].genes.values[0]][0].split(',')\n",
    "            \n",
    "            if not matching_genes_list : continue\n",
    "            \n",
    "            matching_genes_set =all_genes_set.intersection(matching_genes_list)\n",
    "            \n",
    "            if not matching_genes_set : continue\n",
    "            #np.random.seed(seed=1111)\n",
    "            \n",
    "            matching_genes_set_to_select=matching_genes_set.difference(matched_genes_tracker)\n",
    "            \n",
    "            matching_genes_set_to_select=matching_genes_set_to_select if len(matching_genes_set_to_select)>=1 else matching_genes_set\n",
    "            \n",
    "            matching_gene = np.random.choice(list(matching_genes_set_to_select), 1)[0]\n",
    "            \n",
    "            genes.append(temp_gn)\n",
    "            \n",
    "            matching_genes.append(matching_gene)\n",
    "            \n",
    "            matching_genes_dict[temp_gn] = matching_gene\n",
    "            \n",
    "            matched_genes_tracker.add(matching_gene)\n",
    "            \n",
    "        del matching_genes_list\n",
    "        \n",
    "        del temp_gene_modification_profile_df\n",
    "            \n",
    "        del genes_to_select_from_set\n",
    "            \n",
    "        del matching_genes_dict\n",
    "            \n",
    "        del genes_to_select_from_temp_list\n",
    "            \n",
    "        del checked_genes\n",
    "        \n",
    "        del matched_genes_tracker\n",
    "        inputdata.loc[genes,cells]=np.array(inputdata.loc[matching_genes,np.random.choice(all_cells,len_cells,replace=False)])\n",
    "        out_dict['genes'] = genes\n",
    "        out_dict['cells'] = cells\n",
    "        out_dict['fc'] = fc\n",
    "        out_dict['data'] = inputdata\n",
    "        out_dict['matching_genes']=matching_genes\n",
    "    return out_dict\n",
    "\n",
    "\n",
    "def modification_in_subspace_for_selected_genes(inputdata,cells,target_genes,fc,gene_parameters_df,all_genes_set=None,\n",
    "                                                genes_to_select_from_set=None,gene_modification_profile_df=None):\n",
    "    \n",
    "    '''Given specific cells and genes, modifys the expression levels based on the provided fold change and\n",
    "    ensures the mean is within a target subspace '''\n",
    "    all_genes_set=set(inputdata.index.tolist()) if all_genes_set is None else all_genes_set.intersection(inputdata.index.tolist())\n",
    "    genes_to_select_from_set=all_genes_set if genes_to_select_from_set is None else genes_to_select_from_set.intersection(all_genes_set)\n",
    "    gene_parameters_df=gene_parameters_df.loc[all_genes_set,:]\n",
    "    len_cells=len(cells)\n",
    "    all_cells=inputdata.columns.tolist()     \n",
    "    genes=[] \n",
    "    no_genes=len(target_genes)\n",
    "    matching_genes=[]\n",
    "    matching_genes_dict={}\n",
    "    out_dict={}\n",
    "    profiled_fc=None if gene_modification_profile_df is None else gene_modification_profile_df.fc.unique().tolist()\n",
    "    out_dict={}\n",
    "    if fc==0.0:\n",
    "        out_dict['genes'] = target_genes\n",
    "        out_dict['cells'] = cells\n",
    "        out_dict['fc'] = fc\n",
    "        out_dict['data'] = inputdata\n",
    "        out_dict['matching_genes']=target_genes  \n",
    "    elif profiled_fc is None or fc not in profiled_fc:\n",
    "        mean_series = gene_parameters_df.mean_expr\n",
    "        genes_to_select_from_temp_list=list(genes_to_select_from_set)\n",
    "        len_genes_to_select_from_temp_list=len(genes_to_select_from_temp_list)\n",
    "        len_limit=no_genes if no_genes<=len_genes_to_select_from_temp_list else len_genes_to_select_from_temp_list\n",
    "        matched_genes_tracker=set()\n",
    "        for temp_gn in target_genes:\n",
    "            target_mean = np.float(mean_series.loc[temp_gn])*fc\n",
    "            interval=0.1*target_mean \n",
    "            mod_mean_interval = [target_mean - interval, target_mean+ interval]\n",
    "            temp_matching_genes_list = gene_parameters_df[(gene_parameters_df.mean_expr >= mod_mean_interval[0]) &\n",
    "                                                          (gene_parameters_df.mean_expr <= mod_mean_interval[1])].index.tolist()\n",
    "            if not temp_matching_genes_list: continue\n",
    "            matching_genes_list=list(all_genes_set.intersection(temp_matching_genes_list))\n",
    "            if not matching_genes_list : continue\n",
    "            matching_genes_set=set(matching_genes_list).difference(matched_genes_tracker)\n",
    "            matching_gene=np.random.choice(list(matching_genes_set),1)[0] if len(matching_genes_set)>=1 else np.random.choice(matching_genes_list,1)[0]\n",
    "            #if not matching_gene: continue\n",
    "            genes.append(temp_gn)\n",
    "            matching_genes.append(matching_gene)\n",
    "            matching_genes_dict[temp_gn]=matching_gene\n",
    "            matched_genes_tracker.add(matching_gene)\n",
    "        del matching_genes_dict\n",
    "        del genes_to_select_from_set\n",
    "        del mean_series  \n",
    "        del all_genes_set \n",
    "        del gene_parameters_df  \n",
    "        del profiled_fc   \n",
    "        del matched_genes_tracker\n",
    "        inputdata.loc[genes,cells]=np.array(inputdata.loc[matching_genes,\n",
    "                                                          np.random.choice(all_cells,len_cells,replace=False)])\n",
    "        out_dict['genes'] = genes\n",
    "        out_dict['cells'] = cells\n",
    "        out_dict['fc'] = fc\n",
    "        out_dict['data'] = inputdata\n",
    "        out_dict['matching_genes']=matching_genes\n",
    "                         \n",
    "    else:\n",
    "        gene_modification_profile_df=gene_modification_profile_df.astype({'genes':str})\n",
    "        temp_gene_modification_profile_df = gene_modification_profile_df[(gene_modification_profile_df.fc == fc)\n",
    "                                                                         & (gene_modification_profile_df.genes!='nan')]\n",
    "        genes_to_select_from_set = genes_to_select_from_set.intersection(temp_gene_modification_profile_df.gene.tolist())\n",
    "        genes_to_select_from_temp_list=list(genes_to_select_from_set)\n",
    "        len_genes_to_select_from_temp_list=len(genes_to_select_from_temp_list)\n",
    "        len_limit=no_genes if no_genes<=len_genes_to_select_from_temp_list else len_genes_to_select_from_temp_list\n",
    "        matched_genes_tracker=set()\n",
    "        for temp_gn in target_genes:\n",
    "            matching_genes_list = [temp_gene_modification_profile_df[temp_gene_modification_profile_df.gene == temp_gn].genes.values[0]][0].split(',')\n",
    "            if not matching_genes_list : continue\n",
    "            matching_genes_set =all_genes_set.intersection(matching_genes_list)\n",
    "            if not matching_genes_set : continue\n",
    "            print( matching_genes_set)\n",
    "            matching_genes_set_to_select=matching_genes_set.difference(matched_genes_tracker)\n",
    "            matching_genes_set_to_select=matching_genes_set_to_select if len(matching_genes_set_to_select)>=1 else matching_genes_set\n",
    "            matching_gene = np.random.choice(list(matching_genes_set_to_select), 1)[0]\n",
    "            genes.append(temp_gn)\n",
    "            matching_genes.append(matching_gene)\n",
    "            matched_genes_tracker.add(matching_gene)\n",
    "        \n",
    "        \n",
    "        del matching_genes_list\n",
    "        del temp_gene_modification_profile_df    \n",
    "        del genes_to_select_from_set \n",
    "        del matching_genes_dict \n",
    "        del genes_to_select_from_temp_list\n",
    "        del matched_genes_tracker\n",
    "        inputdata.loc[genes,cells]=np.array(inputdata.loc[matching_genes,\n",
    "                                                          np.random.choice(all_cells,len_cells,replace=False)])\n",
    "        out_dict['genes'] = genes\n",
    "        out_dict['cells'] = cells\n",
    "        out_dict['fc'] = fc\n",
    "        out_dict['data'] = inputdata\n",
    "        out_dict['matching_genes']=matching_genes\n",
    "        print('To confirm if its correct')\n",
    "      \n",
    "    return out_dict\n",
    "\n",
    "def add_marker_gene_modification(inputdata,cells,no_genes,fc,gene_parameters_df,all_genes_set=None,\n",
    "                                 genes_to_select_from_set=None,gene_modification_profile_df=None):\n",
    "    \n",
    "    '''Given specific cells and number of genes, modifys the expression levels based on the provided fold change and\n",
    "    ensures the mean is within a target subspace and concatenates the markers to the expression table'''\n",
    "    \n",
    "    all_genes_set=set(inputdata.index.tolist()) if all_genes_set is None else all_genes_set.intersection(inputdata.index.tolist())\n",
    "    profiled_fc=None if gene_modification_profile_df is None else gene_modification_profile_df.fc.unique().tolist()\n",
    "    genes_to_select_from_set=all_genes_set if genes_to_select_from_set is None else genes_to_select_from_set.intersection(all_genes_set)\n",
    "    gene_parameters_df=gene_parameters_df.loc[all_genes_set,:]\n",
    "    len_cells=len(cells)\n",
    "    all_cells=inputdata.columns.tolist() \n",
    "    non_selected_cells=set(all_cells).difference(cells)\n",
    "    genes=[]      \n",
    "    matching_genes=[]\n",
    "    matching_genes_dict={}\n",
    "    out_dict={}\n",
    "    if fc==0.0:\n",
    "        genes_to_select_from_temp_list=list(genes_to_select_from_set)\n",
    "        selected_genes=np.random.choice(genes_to_select_from_temp_list,no_genes,replace=False).tolist()\n",
    "        out_dict['genes'] = selected_genes\n",
    "        out_dict['cells'] = cells\n",
    "        out_dict['fc'] = fc\n",
    "        out_dict['data'] = inputdata\n",
    "        out_dict['matching_genes']=selected_genes \n",
    "    elif profiled_fc is None or fc not in profiled_fc:\n",
    "        mean_series = gene_parameters_df.mean_expr\n",
    "        genes_to_select_from_temp_list=list(genes_to_select_from_set)\n",
    "        len_genes_to_select_from_temp_list=len(genes_to_select_from_temp_list)\n",
    "        len_limit=no_genes if no_genes<=len_genes_to_select_from_temp_list else len_genes_to_select_from_temp_list\n",
    "        checked_genes = []\n",
    "        matched_genes_tracker=set()\n",
    "        while len(matching_genes_dict)<len_limit:\n",
    "            temp_gns_to_check=list(genes_to_select_from_set.difference(checked_genes))\n",
    "            \n",
    "            if len(temp_gns_to_check)==0:break\n",
    "            \n",
    "            temp_gn=np.random.choice(temp_gns_to_check, 1)[0]\n",
    "            \n",
    "            checked_genes.append(temp_gn)\n",
    "            \n",
    "            target_mean = np.float(mean_series.loc[temp_gn])*fc\n",
    "            \n",
    "            interval=0.1*target_mean\n",
    "            \n",
    "            mod_mean_interval = [target_mean - interval, target_mean+ interval]\n",
    "            \n",
    "            temp_matching_genes_list = gene_parameters_df[(gene_parameters_df.mean_expr >= mod_mean_interval[0]) &\n",
    "                                                          (gene_parameters_df.mean_expr <= mod_mean_interval[1])].index.tolist()\n",
    "            \n",
    "            if not temp_matching_genes_list: continue\n",
    "           \n",
    "            #np.random.seed(seed=1111)\n",
    "            \n",
    "            matching_genes_list=list(all_genes_set.intersection(temp_matching_genes_list))\n",
    "            \n",
    "            if not matching_genes_list : continue\n",
    "                \n",
    "            matching_genes_set=set(matching_genes_list).difference(matched_genes_tracker)\n",
    "            \n",
    "            matching_gene=np.random.choice(list(matching_genes_set),1)[0] if len(matching_genes_set)>=1 else np.random.choice(matching_genes_list,1)[0]\n",
    "            \n",
    "            #if not matching_gene: continue\n",
    "                \n",
    "            genes.append(temp_gn)\n",
    "            \n",
    "            matching_genes.append(matching_gene)\n",
    "            \n",
    "            matching_genes_dict[temp_gn]=matching_gene\n",
    "            \n",
    "            matched_genes_tracker.add(matching_gene)\n",
    "            \n",
    "        del matching_genes_dict\n",
    "        \n",
    "        del genes_to_select_from_set\n",
    "        \n",
    "        del mean_series\n",
    "        \n",
    "        del all_genes_set\n",
    "        \n",
    "        del gene_parameters_df\n",
    "            \n",
    "        del profiled_fc\n",
    "            \n",
    "        del checked_genes \n",
    "        \n",
    "        del matched_genes_tracker\n",
    "        \n",
    "        #inputdata.loc[genes,cells]=np.array(inputdata.loc[matching_genes,np.random.choice(all_cells,len_cells,replace=False)])\n",
    "        \n",
    "        first_sp_array=np.array(inputdata.loc[genes,non_selected_cells])\n",
    "        \n",
    "        sec_sp_array=np.array(inputdata.loc[matching_genes,np.random.choice(all_cells,len_cells,replace=False)])\n",
    "      \n",
    "        first_and_sec_sp_array=np.concatenate([first_sp_array,sec_sp_array],axis=1)\n",
    "      \n",
    "        md_expr_cells=list(non_selected_cells)\n",
    "        md_expr_cells.extend(cells)\n",
    "        temp_mod_df=pd.DataFrame(first_and_sec_sp_array,index=['syn_marker_'+ str(md_gn_ind+1) for \n",
    "                                                               md_gn_ind in range(no_genes) ],\n",
    "                                 columns=md_expr_cells)\n",
    "        inputdata=pd.concat([inputdata,temp_mod_df.loc[:,inputdata.columns]])\n",
    "        \n",
    "        out_dict['genes'] = genes\n",
    "        out_dict['cells'] = cells\n",
    "        out_dict['fc'] = fc\n",
    "        out_dict['data'] = inputdata\n",
    "        out_dict['matching_genes']=matching_genes\n",
    "       \n",
    "      \n",
    "    else:\n",
    "        \n",
    "        gene_modification_profile_df=gene_modification_profile_df.astype({'genes':str})\n",
    "        \n",
    "        temp_gene_modification_profile_df = gene_modification_profile_df[(gene_modification_profile_df.fc == fc)\n",
    "                                                                         & (gene_modification_profile_df.genes!='nan')]\n",
    "        \n",
    "        genes_to_select_from_set = genes_to_select_from_set.intersection(temp_gene_modification_profile_df.gene.tolist())\n",
    "        \n",
    "        genes_to_select_from_temp_list=list(genes_to_select_from_set)\n",
    "        \n",
    "        len_genes_to_select_from_temp_list=len(genes_to_select_from_temp_list)\n",
    "        \n",
    "        len_limit=no_genes if no_genes<=len_genes_to_select_from_temp_list else len_genes_to_select_from_temp_list\n",
    "        \n",
    "        checked_genes=[]\n",
    "        \n",
    "        matched_genes_tracker=set()\n",
    "        \n",
    "        while len(matching_genes_dict) < len_limit:\n",
    "            \n",
    "            temp_gns_to_check=list(genes_to_select_from_set.difference(checked_genes))\n",
    "            \n",
    "            if len(temp_gns_to_check)==0:break\n",
    "            \n",
    "            temp_gn=np.random.choice(temp_gns_to_check, 1)[0]\n",
    "            \n",
    "            checked_genes.append(temp_gn)\n",
    "            \n",
    "            matching_genes_list = [temp_gene_modification_profile_df[temp_gene_modification_profile_df.gene == temp_gn].genes.values[0]][0].split(',')\n",
    "            \n",
    "            if not matching_genes_list : continue\n",
    "            \n",
    "            matching_genes_set =all_genes_set.intersection(matching_genes_list)\n",
    "            \n",
    "            if not matching_genes_set : continue\n",
    "            #np.random.seed(seed=1111)\n",
    "            \n",
    "            matching_genes_set_to_select=matching_genes_set.difference(matched_genes_tracker)\n",
    "            \n",
    "            matching_genes_set_to_select=matching_genes_set_to_select if len(matching_genes_set_to_select)>=1 else matching_genes_set\n",
    "            \n",
    "            matching_gene = np.random.choice(list(matching_genes_set_to_select), 1)[0]\n",
    "            \n",
    "            genes.append(temp_gn)\n",
    "            \n",
    "            matching_genes.append(matching_gene)\n",
    "            \n",
    "            matching_genes_dict[temp_gn] = matching_gene\n",
    "            \n",
    "            matched_genes_tracker.add(matching_gene)\n",
    "            \n",
    "        del matching_genes_list\n",
    "        \n",
    "        del temp_gene_modification_profile_df\n",
    "            \n",
    "        del genes_to_select_from_set\n",
    "            \n",
    "        del matching_genes_dict\n",
    "            \n",
    "        del genes_to_select_from_temp_list\n",
    "            \n",
    "        del checked_genes\n",
    "        \n",
    "        del matched_genes_tracker\n",
    "        \n",
    "        #inputdata.loc[genes,cells]=np.array(inputdata.loc[matching_genes,np.random.choice(all_cells,len_cells,replace=False)])\n",
    "        \n",
    "        first_sp_array=np.array(inputdata.loc[genes,non_selected_cells])\n",
    "        sec_sp_array=np.array(inputdata.loc[matching_genes,np.random.choice(all_cells,len_cells,replace=False)])\n",
    "        first_and_sec_sp_array=np.concatenate([first_sp_array,sec_sp_array],axis=1)\n",
    "        md_expr_cells=list(non_selected_cells)\n",
    "        md_expr_cells.extend(cells)\n",
    "        temp_mod_df=pd.DataFrame(first_and_sec_sp_array,index=['syn_marker_'+ str(md_gn_ind+1) for \n",
    "                                                               md_gn_ind in range(no_genes) ],\n",
    "                                 columns=md_expr_cells)\n",
    "        inputdata=pd.concat([inputdata,temp_mod_df.loc[:,inputdata.columns]])\n",
    "        \n",
    "        out_dict['genes'] = genes\n",
    "        out_dict['cells'] = cells\n",
    "        out_dict['fc'] = fc\n",
    "        out_dict['data'] = inputdata\n",
    "        out_dict['matching_genes']=matching_genes\n",
    "    \n",
    "    return out_dict\n",
    "def mean_modification(inputdata,genes,cells,fc=2):\n",
    "   \n",
    "    ''' Method modifys the mean expression of target genes and cells based on a fold change  \n",
    "    and generates normally distributed expression '''\n",
    "    \n",
    "    inputdata_mean_series=inputdata.mean(axis=1)\n",
    "    inputdata.loc[genes, cells] = [np.random.normal(loc=inputdata_mean_series.loc[gn]*fc,\n",
    "                                           size=len(cells))for gn in genes]\n",
    "    return inputdata\n",
    "\n",
    "def add_synthetic_genes_modification(inputdata,genes,cells,fc=2,std=1.0):\n",
    "    \n",
    "    '''Given specific samples and genes, modifys the expression levels based on their mean expression \n",
    "    by adding synthetic genes provided fold change.'''\n",
    "    \n",
    "    inputdata_mean_series=inputdata.mean(axis=1)\n",
    "    \n",
    "    all_scs=inputdata.columns\n",
    "    \n",
    "    target_genes_df = pd.DataFrame([[random.gauss(mu=inputdata_mean_series.loc[gn]*fc, sigma=std) \n",
    "                                     if sc in cells else random.gauss(mu=inputdata_mean_series.loc[gn], sigma=std)\n",
    "                                     for sc in all_scs] for gn in genes],index=['synthetic_gene_'+str(ind+1) \n",
    "                                                                                             for ind in range(len(genes))],\n",
    "                                  columns=inputdata.columns)\n",
    "    \n",
    "    outputdata=inputdata.append(target_genes_df)\n",
    "    \n",
    "    return outputdata\n",
    "def filter_genes_and_samples(inputdata,cut_off=0,no_samples=1,genes_count_cut_off=1):\n",
    "    \n",
    "    '''Filters genes based on the cut off sample count(s) from a dataframe'''\n",
    "    \n",
    "    cut_off=float(cut_off)\n",
    "    \n",
    "    no_samples=int(no_samples)\n",
    "    \n",
    "    genes_count_cut_off=int(genes_count_cut_off)\n",
    "    \n",
    "    selected_samples=inputdata.columns[inputdata[inputdata>cut_off].count(axis=0)>=genes_count_cut_off]\n",
    "    \n",
    "    local_no_samples=no_samples if len(selected_samples)>=no_samples else len(selected_samples)\n",
    "    \n",
    "    selected_gns=inputdata.index[inputdata[inputdata>cut_off].count(axis=1)>=local_no_samples]\n",
    "    \n",
    "    outputdata=inputdata.loc[selected_gns,selected_samples]\n",
    "    return outputdata\n",
    "def remove_nonexpressedgenes(inputdata): \n",
    "    \n",
    "    '''Filters genes with zero expression in all samples'''\n",
    "    \n",
    "    selected_gns=inputdata.index[inputdata[inputdata>0.0].count(axis=1)>=1]\n",
    "    \n",
    "    selected_samples=inputdata.columns[inputdata[inputdata>0.0].count(axis=0)>=1]\n",
    "    \n",
    "    output_data=inputdata.loc[selected_gns,selected_samples]\n",
    "    \n",
    "    return output_data\n",
    "def remove_genes(inputdata, gene_pattern):\n",
    "    \n",
    "    '''Removes genes whose ids or names mateches the provided pattern. Good for removing e.g ERCCs'''\n",
    "    \n",
    "    pattern=re.compile(gene_pattern)\n",
    "    \n",
    "    genes_to_keep=[gn for gn in inputdata.index if pattern.search(string=gn)==None]\n",
    "   \n",
    "    outdata=inputdata.loc[genes_to_keep,:]\n",
    "    \n",
    "    return outdata\n",
    "\n",
    "def run_kmeans(input_array,no_clusters=2,scale_array=True):\n",
    "    \n",
    "    ''' Runs K-means clustering on the input array'''\n",
    "    #np.random.seed(2573780)\n",
    "    data = scale(input_array) if scale_array else input_array\n",
    "    kmean = KMeans(init='k-means++', n_clusters=no_clusters, n_init=10).fit(data)\n",
    "    return kmean\n",
    "def run_knn(input_array,labels,no_neighbours=5, radius=1.0,algorithm='auto',leaf_size=30,distance_metric='minkowski',\n",
    "            p=2,cpus=1):\n",
    "    \n",
    "    '''Run the k-nearest neighbours for classification'''\n",
    "   \n",
    "    input_array=np.array(input_array)\n",
    "    \n",
    "    clf = neighbors.KNeighborsClassifier(n_neighbors=no_neighbours,radius=radius,algorithm=algorithm,\n",
    "                                         leaf_size=leaf_size,metric=distance_metric,p=p,n_jobs=cpus)\n",
    "    clf.fit(input_array, labels)\n",
    "    \n",
    "    prediction_res = clf.predict(input_array)\n",
    "  \n",
    "    return(prediction_res)\n",
    "def run_dbscan(input_array,in_min_cluster_size=5,in_min_samples=None, in_metric='euclidean'):\n",
    "    ''' Runs hdbscan clustering on the input array'''\n",
    "    #np.random.seed(2573780)\n",
    "    hdbscan_clusterer = hdbscan.HDBSCAN(min_cluster_size=in_min_cluster_size,min_samples=in_min_samples,\n",
    "                                metric=in_metric).fit(input_array)\n",
    "    return hdbscan_clusterer\n",
    "\n",
    "def get_homogeneity_score(input_array,labels,clustering_method='knn',knn_neighbours=3,\n",
    "                          knn_radius=1.0,knn_algorithm='auto',knn_leaf_size=30,knn_distance_metric='minkowski',\n",
    "                          knn_p=2,knn_cpus=1,kmeans_clusters=2,scale_kmeans=True,dbscan_min_cluster_size=5,\n",
    "                          dbscan_min_samples=None,dbscan_metric='euclidean'):\n",
    "    '''Runs the input data and generates the homogeneity score'''\n",
    "    if clustering_method=='knn':\n",
    "        knn_labels=run_knn(input_array=input_array,labels=labels,no_neighbours=knn_neighbours,\n",
    "                           radius=knn_radius,algorithm=knn_algorithm,leaf_size=knn_leaf_size,\n",
    "                           distance_metric=knn_distance_metric,p=knn_p,cpus=knn_cpus)\n",
    "        hs = metrics.homogeneity_score(labels_true=labels,labels_pred=knn_labels)\n",
    "        complet_score=metrics.completeness_score(labels_true=labels, labels_pred=knn_labels) \n",
    "        v_score=metrics.v_measure_score(labels_true=labels, labels_pred=knn_labels)\n",
    "        adj_rand_index=metrics.adjusted_rand_score(labels_true=labels, labels_pred=knn_labels)\n",
    "        adj_mutual_info=metrics.adjusted_mutual_info_score(labels_true=labels, labels_pred=knn_labels)\n",
    "        fm_score=metrics.fowlkes_mallows_score(labels_true=labels, labels_pred=knn_labels)\n",
    "        euclidean_silhouette_score=metrics.silhouette_score(X=input_array, labels=knn_labels, metric='euclidean')\n",
    "        cosine_silhouette_score=metrics.silhouette_score(X=input_array, labels=knn_labels, metric='cosine')\n",
    "        calinski_harabaz_score=metrics.calinski_harabaz_score(X=input_array, labels=knn_labels)\n",
    "        return {'hs':hs,'labs':knn_labels,'complet_score':complet_score,'v_score':v_score,'adj_rand_index':adj_rand_index,\n",
    "               'adj_mutual_info':adj_mutual_info,'fm_score':fm_score,'calinski_harabaz_score':calinski_harabaz_score,\n",
    "               'cosine_silhouette_score':cosine_silhouette_score,'euclidean_silhouette_score':euclidean_silhouette_score}\n",
    "    \n",
    "    if clustering_method=='kmeans':\n",
    "        kmeans_result=run_kmeans(input_array=input_array,no_clusters=kmeans_clusters,scale_array=scale_kmeans)\n",
    "        hs = metrics.homogeneity_score(labels_true=labels,labels_pred=kmeans_result.labels_)\n",
    "        complet_score=metrics.completeness_score(labels_true=labels, labels_pred=kmeans_result.labels_) \n",
    "        v_score=metrics.v_measure_score(labels_true=labels, labels_pred=kmeans_result.labels_)\n",
    "        adj_rand_index=metrics.adjusted_rand_score(labels_true=labels, labels_pred=kmeans_result.labels_)\n",
    "        adj_mutual_info=metrics.adjusted_mutual_info_score(labels_true=labels, labels_pred=kmeans_result.labels_,\n",
    "                                                           average_method='arithmetic')\n",
    "        fm_score=metrics.fowlkes_mallows_score(labels_true=labels, labels_pred=kmeans_result.labels_)\n",
    "        euclidean_silhouette_score=metrics.silhouette_score(X=input_array, labels=kmeans_result.labels_, metric='euclidean')\n",
    "        cosine_silhouette_score=metrics.silhouette_score(X=input_array, labels=kmeans_result.labels_, metric='cosine')\n",
    "        calinski_harabaz_score=metrics.calinski_harabaz_score(X=input_array, labels=kmeans_result.labels_)\n",
    "        return {'hs':hs,'labs':kmeans_result.labels_,'complet_score':complet_score,'v_score':v_score,\n",
    "                'adj_rand_index':adj_rand_index,'adj_mutual_info':adj_mutual_info,'fm_score':fm_score,\n",
    "                'calinski_harabaz_score':calinski_harabaz_score,'cosine_silhouette_score':cosine_silhouette_score,\n",
    "                'euclidean_silhouette_score':euclidean_silhouette_score}\n",
    "    \n",
    "    if clustering_method=='hdbscan':\n",
    "        hdbscan_result=run_dbscan(input_array=input_array,in_min_cluster_size=dbscan_min_cluster_size,\n",
    "                                  in_min_samples=dbscan_min_samples, in_metric=dbscan_metric)\n",
    "        hs = metrics.homogeneity_score(labels_true=labels,labels_pred=hdbscan_result.labels_)\n",
    "        complet_score=metrics.completeness_score(labels_true=labels, labels_pred=hdbscan_result.labels_) \n",
    "        v_score=metrics.v_measure_score(labels_true=labels, labels_pred=hdbscan_result.labels_)\n",
    "        adj_rand_index=metrics.adjusted_rand_score(labels_true=labels, labels_pred=hdbscan_result.labels_)\n",
    "        adj_mutual_info=metrics.adjusted_mutual_info_score(labels_true=labels, labels_pred=hdbscan_result.labels_,\n",
    "                                                           average_method='arithmetic')\n",
    "        fm_score=metrics.fowlkes_mallows_score(labels_true=labels, labels_pred=hdbscan_result.labels_)\n",
    "        euclidean_silhouette_score=metrics.silhouette_score(X=input_array, labels=hdbscan_result.labels_, metric='euclidean')\n",
    "        cosine_silhouette_score=metrics.silhouette_score(X=input_array, labels=hdbscan_result.labels_, metric='cosine')\n",
    "        calinski_harabaz_score=metrics.calinski_harabaz_score(X=input_array, labels=hdbscan_result.labels_)\n",
    "        #hs =0.0 if hs <0 else hs \n",
    "        return {'hs':hs,'labs':hdbscan_result.labels_,'complet_score':complet_score,'v_score':v_score,\n",
    "                'adj_rand_index':adj_rand_index,'adj_mutual_info':adj_mutual_info,'fm_score':fm_score,\n",
    "                'calinski_harabaz_score':calinski_harabaz_score,'cosine_silhouette_score':cosine_silhouette_score,\n",
    "                'euclidean_silhouette_score':euclidean_silhouette_score}\n",
    "    \n",
    "    pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>method</th>\n",
       "      <th>hs</th>\n",
       "      <th>complet_score</th>\n",
       "      <th>v_score</th>\n",
       "      <th>adj_rand_index</th>\n",
       "      <th>adj_mutual_info</th>\n",
       "      <th>fm_score</th>\n",
       "      <th>calinski_harabaz_score</th>\n",
       "      <th>cosine_silhouette_score</th>\n",
       "      <th>euclidean_silhouette_score</th>\n",
       "      <th>...</th>\n",
       "      <th>bin_start</th>\n",
       "      <th>bin_end</th>\n",
       "      <th>protocol</th>\n",
       "      <th>meanTopExpressedGenes</th>\n",
       "      <th>nTopGenes</th>\n",
       "      <th>post_mod_genes</th>\n",
       "      <th>genes_in_all_cells</th>\n",
       "      <th>spike_ins_pattern</th>\n",
       "      <th>normalization_method</th>\n",
       "      <th>nTopVarRemovedGenes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>pca_tsne</td>\n",
       "      <td>0.007299</td>\n",
       "      <td>0.007352</td>\n",
       "      <td>0.007326</td>\n",
       "      <td>-0.044420</td>\n",
       "      <td>-0.033232</td>\n",
       "      <td>0.453046</td>\n",
       "      <td>11.878872</td>\n",
       "      <td>0.508358</td>\n",
       "      <td>0.282807</td>\n",
       "      <td>...</td>\n",
       "      <td>5000</td>\n",
       "      <td>7500</td>\n",
       "      <td>Smartseq2</td>\n",
       "      <td>10000</td>\n",
       "      <td>500</td>\n",
       "      <td>sig_only</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>scran</td>\n",
       "      <td>500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>pca_tsne</td>\n",
       "      <td>0.295807</td>\n",
       "      <td>0.304657</td>\n",
       "      <td>0.300167</td>\n",
       "      <td>0.325942</td>\n",
       "      <td>0.271097</td>\n",
       "      <td>0.652328</td>\n",
       "      <td>0.928676</td>\n",
       "      <td>0.047435</td>\n",
       "      <td>0.062614</td>\n",
       "      <td>...</td>\n",
       "      <td>5000</td>\n",
       "      <td>7500</td>\n",
       "      <td>Smartseq2</td>\n",
       "      <td>10000</td>\n",
       "      <td>500</td>\n",
       "      <td>sig_only</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>scran</td>\n",
       "      <td>500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>pca_tsne</td>\n",
       "      <td>0.397313</td>\n",
       "      <td>0.400204</td>\n",
       "      <td>0.398753</td>\n",
       "      <td>0.461966</td>\n",
       "      <td>0.374188</td>\n",
       "      <td>0.718243</td>\n",
       "      <td>13.663188</td>\n",
       "      <td>0.580842</td>\n",
       "      <td>0.318776</td>\n",
       "      <td>...</td>\n",
       "      <td>5000</td>\n",
       "      <td>7500</td>\n",
       "      <td>Smartseq2</td>\n",
       "      <td>10000</td>\n",
       "      <td>500</td>\n",
       "      <td>sig_only</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>scran</td>\n",
       "      <td>500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>pca_tsne</td>\n",
       "      <td>0.397313</td>\n",
       "      <td>0.400204</td>\n",
       "      <td>0.398753</td>\n",
       "      <td>0.461966</td>\n",
       "      <td>0.374188</td>\n",
       "      <td>0.718243</td>\n",
       "      <td>7.274006</td>\n",
       "      <td>0.422801</td>\n",
       "      <td>0.209705</td>\n",
       "      <td>...</td>\n",
       "      <td>5000</td>\n",
       "      <td>7500</td>\n",
       "      <td>Smartseq2</td>\n",
       "      <td>10000</td>\n",
       "      <td>500</td>\n",
       "      <td>sig_only</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>scran</td>\n",
       "      <td>500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>pca_tsne</td>\n",
       "      <td>0.118709</td>\n",
       "      <td>0.118709</td>\n",
       "      <td>0.118709</td>\n",
       "      <td>0.113333</td>\n",
       "      <td>0.082893</td>\n",
       "      <td>0.533333</td>\n",
       "      <td>2.340706</td>\n",
       "      <td>0.132010</td>\n",
       "      <td>0.092911</td>\n",
       "      <td>...</td>\n",
       "      <td>5000</td>\n",
       "      <td>7500</td>\n",
       "      <td>Smartseq2</td>\n",
       "      <td>10000</td>\n",
       "      <td>500</td>\n",
       "      <td>sig_only</td>\n",
       "      <td>False</td>\n",
       "      <td>NaN</td>\n",
       "      <td>scran</td>\n",
       "      <td>500</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows  55 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     method        hs  complet_score   v_score  adj_rand_index  \\\n",
       "0  pca_tsne  0.007299       0.007352  0.007326       -0.044420   \n",
       "1  pca_tsne  0.295807       0.304657  0.300167        0.325942   \n",
       "0  pca_tsne  0.397313       0.400204  0.398753        0.461966   \n",
       "1  pca_tsne  0.397313       0.400204  0.398753        0.461966   \n",
       "0  pca_tsne  0.118709       0.118709  0.118709        0.113333   \n",
       "\n",
       "   adj_mutual_info  fm_score  calinski_harabaz_score  cosine_silhouette_score  \\\n",
       "0        -0.033232  0.453046               11.878872                 0.508358   \n",
       "1         0.271097  0.652328                0.928676                 0.047435   \n",
       "0         0.374188  0.718243               13.663188                 0.580842   \n",
       "1         0.374188  0.718243                7.274006                 0.422801   \n",
       "0         0.082893  0.533333                2.340706                 0.132010   \n",
       "\n",
       "   euclidean_silhouette_score  ...  bin_start  bin_end   protocol  \\\n",
       "0                    0.282807  ...       5000     7500  Smartseq2   \n",
       "1                    0.062614  ...       5000     7500  Smartseq2   \n",
       "0                    0.318776  ...       5000     7500  Smartseq2   \n",
       "1                    0.209705  ...       5000     7500  Smartseq2   \n",
       "0                    0.092911  ...       5000     7500  Smartseq2   \n",
       "\n",
       "   meanTopExpressedGenes  nTopGenes  post_mod_genes  genes_in_all_cells  \\\n",
       "0                  10000        500        sig_only               False   \n",
       "1                  10000        500        sig_only               False   \n",
       "0                  10000        500        sig_only               False   \n",
       "1                  10000        500        sig_only               False   \n",
       "0                  10000        500        sig_only               False   \n",
       "\n",
       "   spike_ins_pattern  normalization_method  nTopVarRemovedGenes  \n",
       "0                NaN                 scran                  500  \n",
       "1                NaN                 scran                  500  \n",
       "0                NaN                 scran                  500  \n",
       "1                NaN                 scran                  500  \n",
       "0                NaN                 scran                  500  \n",
       "\n",
       "[5 rows x 55 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hs_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(127, 55)"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#%%script false\n",
    "test_df=hs_df[hs_df.hs.notna()]\n",
    "split_groups =['cells_count','modification_method', 'clustering_method', 'randomize_cells_labels','method',\n",
    "               'top_var_genes_cut_off','top_var_genes_included', 'pca_components','post_mod_genes','bin_start',\n",
    "               'bin_end','normalization_method']\n",
    "replicates_grps=['cells_count','modification_method', 'clustering_method','fc','genes_count', \n",
    "                 'randomize_cells_labels','method','top_var_genes_cut_off','top_var_genes_included',\n",
    "                 'normalization_method', 'pca_components','bin_start','bin_end','protocol']\n",
    "hue='protocol'\n",
    "#homo_scores_df['genes_count']=homo_scores_df.len_mod_and_var_genes.values\n",
    "replicates_cut_off=90\n",
    "out_dir=None\n",
    "out_dir='results/figures'\n",
    "fc_list=[0.1,  0.2,  0.3,  0.5,  2.0 ,  3.0 ,  5.0 , 10.0]   \n",
    "#fc_list=None\n",
    "#fc_order=[0.2,0.5,2,5]\n",
    "#fc_order=None\n",
    "#x_axis_subset=[10,30,50]\n",
    "x_axis_subset=None\n",
    "in_col_wrap=4\n",
    "show_reps_counts=['fc','genes_count',hue]\n",
    "#show_reps_counts=None\n",
    "hue_order=['Smartseq2', 'CELseq2', 'DropSeq', 'SCRBseq', 'Smartseq', 'MARSseq']\n",
    "#hue_order=None\n",
    "#Globala_variables\n",
    "sns.set_context(context='paper',rc={'style':'white'})\n",
    "#sns.set_palette(palette='colorblind')\n",
    "#palette_dict='bright'\n",
    "x_axis='genes_count'\n",
    "boxprops_dict=dict(linewidth=0)\n",
    "reps_limit=90\n",
    "boxplot_mean=True\n",
    "temp_palette={'SCRBseq': (0.00784313725490196, 0.24313725490196078, 1.0),\n",
    "              'Smartseq2': (1.0, 0.48627450980392156, 0.0),\n",
    "              'CELseq2': (0.10196078431372549, 0.788235294117647, 0.2196078431372549),\n",
    "              'MARSseq': (0.9098039215686274, 0.0, 0.043137254901960784),\n",
    "              'DropSeq': (0.5450980392156862, 0.16862745098039217, 0.8862745098039215),\n",
    "              'Smartseq': (0.6235294117647059, 0.2823529411764706, 0.0)}\n",
    "utils.plot_classification_scores_between_experiments(expriment_path=test_df, out_dir=out_dir,\n",
    "                                                     x_axis=x_axis,split_groups=split_groups,\n",
    "                                                     replicates_grps=replicates_grps,\n",
    "                                                     y_axis='adj_rand_index',in_kind='point',ci=95,\n",
    "                                                     in_col_wrap=in_col_wrap,replicates_per_grp=replicates_cut_off,\n",
    "                                                     col_subset=fc_list,hue=hue,hue_order=hue_order,\n",
    "                                                     x_axis_subset=x_axis_subset,col_order_value=None,\n",
    "                                                     show_reps_counts=show_reps_counts,\n",
    "                                                     palette=temp_palette,facet_size=10,add_swarm=False,\n",
    "                                                     swarm_points_size=5.0,x_axis_rotation=70,\n",
    "                                                     boxplot_mean=boxplot_mean,\n",
    "                                                     boxprops=boxprops_dict,boxplot_outliers=False,\n",
    "                                                     sharex=False,sharey=True,errwidth=0.0001,capsize=0.02, \n",
    "                                                     limit=reps_limit)   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
